File: next/dweb-browser/src/Platforms/iOS/MicroModule/Browser/JsProcess/JsProcessWebApi.cs
using System.Text.Json.Serialization;
using Foundation;

namespace DwebBrowser.MicroService.Browser.JsProcess;

public class JsProcessWebApi : IDisposable
{
    static readonly Debugger Console = new("JsProcessWebApi");
    public DWebView.DWebView DWebView { get; init; }

    public JsProcessWebApi(DWebView.DWebView dWebView)
    {
        DWebView = dWebView;
    }


    public class ProcessInfo
    {
        [JsonPropertyName("process_id")]
        public int ProcessId { get; set; }
        public ProcessInfo(int process_id)
        {
            ProcessId = process_id;
        }
    }

    public record ProcessHandler(ProcessInfo Info, MessagePortIpc Ipc);

    /**
     * 执行js"多步骤"代码时的并发编号
     */
    private int _hidAcc = 0;

    public async Task<ProcessHandler> CreateProcess(
        string env_script_url,
        string metadata_json,
        string env_json,
        Ipc.IMicroModuleInfo remoteModule,
        string host)
    {
        var channel = await DWebView.CreateWebMessageChannel();
        var port1 = channel.Port1;
        var port2 = channel.Port2;

        var hid = Interlocked.Increment(ref _hidAcc);
        var nsProcessInfo = await DWebView.EvaluateAsyncJavascriptCode($$"""
           new Promise((resolve,reject)=>{
                addEventListener("message", async function doCreateProcess(event) {
                    if (event.data === "js-process/create-process/{{hid}}") {
                        removeEventListener("message", doCreateProcess);
                        const fetch_port = event.ports[0];
                        try{
                            resolve(await createProcess(`{{env_script_url}}`,JSON.stringify({{metadata_json}}),JSON.stringify({{env_json}}),fetch_port,`{{host}}`))
                        }catch(err){
                            reject(err)
                        }
                    }
                })
            })
        """.Trim(), () => DWebView.PostMessage("js-process/create-process/" + hid, new[] { port1 }));

        Console.Log("CreateProcess", "processInfo {0}", nsProcessInfo);

        var processId = (int)(NSNumber)nsProcessInfo.ValueForKey(new NSString("process_id"));
        var processInfo = new ProcessInfo(processId);

        return new ProcessHandler(processInfo, new MessagePortIpc(port2, remoteModule, IPC_ROLE.CLIENT));
    }

    public record RunProcessMainOptions(string MainUrl);

    public Task RunProcessMain(int process_id, RunProcessMainOptions options) =>
        DWebView.EvaluateJavaScriptAsync("void runProcessMain(" + process_id + ", {main_url:`" + options.MainUrl + "`})").NoThrow();

    public Task DestroyProcess(int process_id) =>
        MainThread.InvokeOnMainThreadAsync(() =>
        {
            return DWebView.EvaluateJavaScriptAsync(string.Format("void destroyProcess({0})", process_id).Trim()).NoThrow();
        });


    public async Task<int> CreateIpc(int process_id, Mmid mmid)
    {
        var channel = await DWebView.CreateWebMessageChannel();
        var port1 = channel.Port1;
        var port2 = channel.Port2;

        var hid = Interlocked.Increment(ref _hidAcc);
        await DWebView.EvaluateAsyncJavascriptCode($$"""
        new Promise((resolve,reject)=>{
            addEventListener("message", async function doCreateIpc(event) {
                if (event.data === "js-process/create-ipc/{{hid}}") {
                    removeEventListener("message", doCreateIpc);
                    const ipc_port = event.ports[0];
                    try{
                        resolve(await createIpc({{process_id}}, `{{mmid}}`, ipc_port))
                    }catch(err){
                        reject(err)
                    }
                }
            })
        })
        """.Trim(), () => DWebView.PostMessage("js-process/create-ipc/" + hid, new[] { port1 }));

        return IpcWebMessageCache.SaveNative2JsIpcPort(port2);
    }

    public Task CreateIpcFail(int process_id, Mmid mmid, string reason) =>
        MainThread.InvokeOnMainThreadAsync(() =>
        {
            return DWebView.EvaluateJavaScriptAsync(
                string.Format("void createIpcFail({0}, '{1}', '{2}')", process_id, mmid, reason).Trim()).NoThrow();
        });

    public void Dispose()
    {
        //throw new NotImplementedException();
    }
}

File: android/app/src/main/java/info/bagen/dwebbrowser/microService/browser/jsProcess/JsProcessWebApi.kt
package info.bagen.dwebbrowser.microService.browser.jsProcess

import android.net.Uri
import android.webkit.WebMessage
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import org.dweb_browser.dwebview.DWebView
import org.dweb_browser.dwebview.ipcWeb.MessagePortIpc
import org.dweb_browser.dwebview.ipcWeb.saveNative2JsIpcPort
import org.dweb_browser.microservice.help.MMID
import org.dweb_browser.microservice.help.MicroModuleManifest
import org.dweb_browser.microservice.help.gson
import org.dweb_browser.microservice.ipc.helper.IPC_ROLE
import java.util.concurrent.atomic.AtomicInteger

class JsProcessWebApi(val dWebView: DWebView) {
  suspend fun isReady() =
    dWebView.evaluateSyncJavascriptCode("typeof createProcess") == "function"

  data class ProcessInfo(val process_id: Int) {}
  inner class ProcessHandler(val info: ProcessInfo, var ipc: MessagePortIpc)

  /**
   * 执行js"多步骤"代码时的并发编号
   */
  private var hidAcc = AtomicInteger(1);

  suspend fun createProcess(
    env_script_url: String,
    metadata_json: String,
    env_json: String,
    remoteModule: MicroModuleManifest,
    host: String
  ) = withContext(Dispatchers.Main) {
    val channel = dWebView.createWebMessageChannel()
    val port1 = channel[0]
    val port2 = channel[1]
    val metadata_json_str = gson.toJson(metadata_json)
    val env_json_str = gson.toJson(env_json)

    val hid = hidAcc.addAndGet(1);
    val processInfo_json = dWebView.evaluateAsyncJavascriptCode("""
            new Promise((resolve,reject)=>{
                addEventListener("message", async function doCreateProcess(event) {
                    if (event.data === "js-process/create-process/$hid") {
                     try{
                        removeEventListener("message", doCreateProcess);
                        const fetch_port = event.ports[0];
                        resolve(await createProcess(`$env_script_url`,$metadata_json_str,$env_json_str,fetch_port,`$host`))
                        }catch(err){
                            reject(err)
                        }
                    }
                })
            })
            """.trimIndent(), afterEval = {
      try {
        dWebView.postWebMessage(
          WebMessage("js-process/create-process/$hid", arrayOf(port1)), Uri.EMPTY
        );
      } catch (e: Exception) {
        e.printStackTrace()
      }
    })
    debugJsProcess("processInfo", processInfo_json)
    val info = gson.fromJson(processInfo_json, ProcessInfo::class.java)
    val ipc = MessagePortIpc.from(port2, remoteModule, IPC_ROLE.CLIENT)
    ProcessHandler(info, ipc)
  }

  suspend fun createIpcFail(
    process_id: String,
    mmid: String,
    reason: String
  ) = dWebView.evaluateAsyncJavascriptCode(
    """
        createIpcFail($process_id,$mmid,$reason)
        """.trimIndent()
  ).let {}

  data class RunProcessMainOptions(val main_url: String)

  suspend fun runProcessMain(process_id: Int, options: RunProcessMainOptions) =
    dWebView.evaluateAsyncJavascriptCode(
      """
        runProcessMain($process_id, { main_url:`${options.main_url}` })
        """.trimIndent()
    ).let {}

  suspend fun destroyProcess(process_id: Int) =
    dWebView.evaluateAsyncJavascriptCode(
      """
        destroyProcess($process_id)
        """.trimIndent()
    ).let {}

  suspend fun createIpc(process_id: Int, mmid: MMID) = withContext(Dispatchers.Main) {
    val channel = dWebView.createWebMessageChannel()
    val port1 = channel[0]
    val port2 = channel[1]
    val jsIpcPortId = saveNative2JsIpcPort(port2)
    val hid = hidAcc.getAndAdd(1);
    dWebView.evaluateAsyncJavascriptCode("""
        new Promise((resolve,reject)=>{
            addEventListener("message", async function doCreateIpc(event) {
                if (event.data === "js-process/create-ipc/$hid") {
                  try{
                    removeEventListener("message", doCreateIpc);
                    const ipc_port = event.ports[0];
                    resolve(await createIpc($process_id, `$mmid`, ipc_port))
                    }catch(err){
                        reject(err)
                    }
                }
            })
        })
        """.trimIndent(), afterEval = {
      dWebView.postWebMessage(
        WebMessage("js-process/create-ipc/$hid", arrayOf(port1)), Uri.EMPTY
      );
    })
    jsIpcPortId
  }


  fun destroy() {
    dWebView.destroy()
  }
}